File: java/src/test/java/deltix/dfp/Decimal64Test.java
package deltix.dfp;

import org.junit.Assert;
import org.junit.Test;

import java.util.Objects;

public class Decimal64Test {
    @Test
    public void unbox() {
        Assert.assertEquals(5L, Decimal64.toUnderlying(Decimal64.fromUnderlying(5L)));
        Assert.assertEquals(Decimal64Utils.NULL, Decimal64.toUnderlying(null));
    }

    @Test
    public void equality() {

        Decimal64 a = Decimal64.fromDouble(12345.6789);
        Decimal64 a2 = Decimal64.fromDouble(12345.6789);
        Decimal64 b = Decimal64.fromDouble(42);
        String s = "12345.6789";

        Assert.assertTrue("null == null", Decimal64.equals(null, (Object) null));
        Assert.assertTrue("null == null", Decimal64.equals(null, (Decimal64) null));
        Assert.assertTrue("null == null", Decimal64.isIdentical(null, (Object) null));
        Assert.assertTrue("null == null", Decimal64.isIdentical(null, (Decimal64) null));


        Assert.assertFalse("null != a", Decimal64.equals(null, (Object) a));
        Assert.assertFalse("null != a", Decimal64.equals(null, (Decimal64) a));
        Assert.assertFalse("null != a", Decimal64.isIdentical(null, (Object) a));
        Assert.assertFalse("null != a", Decimal64.isIdentical(null, (Decimal64) a));



        Assert.assertFalse("null != s", Decimal64.equals(null, (Object) s));
        Assert.assertFalse("null != s", Decimal64.equals(null, s));
        Assert.assertFalse("null != s", Decimal64.isIdentical(null, s));
        Assert.assertFalse("null != s", Decimal64.isIdentical(null, (Object)s));


        Assert.assertFalse("a != null ", Decimal64.equals(a, null));
        Assert.assertFalse("a != null ", Decimal64.isIdentical(a, null));

        Assert.assertFalse("a != null ", a.equals(null));
        Assert.assertFalse("a != null ", a.isIdentical(null));

        Assert.assertFalse("a != string", Decimal64.equals(a, s));
        Assert.assertFalse("a != string", Decimal64.isIdentical(a, s));
        Assert.assertFalse("a != string", a.equals(s));
        Assert.assertFalse("a != string", a.isIdentical(s));

        Assert.assertTrue("a == a", a == a);
        Assert.assertTrue("a == a", Decimal64.equals(a, a));
        Assert.assertTrue("a == a", Decimal64.isIdentical(a, a));
        Assert.assertTrue("a == a2", Decimal64.equals(a, a2));
        Assert.assertTrue("a == a2", Decimal64.isIdentical(a, a2));

            // Note: these comparisons may be of interest for Value Type Agent
        Assert.assertTrue("a != b", a != b);
        Assert.assertTrue("a != b", (Object)a != (Object)b);
        Assert.assertTrue("a != s", (Object)a != (Object) s);

        Assert.assertFalse("a != b", Decimal64.equals(a, b));
        Assert.assertFalse("a != b", Decimal64.isIdentical(a, b));

        Assert.assertTrue("+NaN == -NaN", Decimal64.equals(Decimal64.NaN, Decimal64.NaN.negate()));
        Assert.assertTrue("+NaN == -NaN", Decimal64.NaN.equals(Decimal64.NaN.negate()));
        Assert.assertTrue("+NaN == -NaN", Decimal64.equals(Decimal64.NaN.negate(), Decimal64.NaN));
        Assert.assertTrue("+NaN == -NaN", Decimal64.NaN.negate().equals(Decimal64.NaN));
        Assert.assertFalse("+Nan != -NaN", Decimal64.isIdentical(Decimal64.NaN, Decimal64.NaN.negate()));
        Assert.assertFalse("+NaN != -NaN", Decimal64.NaN.isIdentical(Decimal64.NaN.negate()));
        Assert.assertFalse("+Nan != -NaN", Decimal64.isIdentical(Decimal64.NaN.negate(), Decimal64.NaN));
        Assert.assertFalse("+NaN != -NaN", Decimal64.NaN.negate().isIdentical(Decimal64.NaN));

        // Equality-only tests for compareTo()
        Assert.assertFalse("+NaN != +NaN", 0 != Decimal64.NaN.compareTo(Decimal64.NaN));
        Assert.assertFalse("+NaN != -NaN", 0 != Decimal64.NaN.compareTo(Decimal64.NaN.negate()));
        Assert.assertFalse("+NaN != -NaN", 0 != Decimal64.NaN.negate().compareTo(Decimal64.NaN.negate()));

        Assert.assertTrue("a == a", 0 == a.compareTo(a));
        Assert.assertTrue("a == a2", 0 == a.compareTo(a2));
        Assert.assertFalse("a != b", 0 == a.compareTo(b));
        // TODO: Check if compareTo throws
    }

    @Test
    public void canonizeTest() {
        long al = Decimal64Utils.fromFixedPoint(10000, 4);
        long bl = Decimal64Utils.fromFixedPoint(1, 0);
        long cl = Decimal64Utils.fromFixedPoint(10, 1);
        long dl = Decimal64Utils.fromFixedPoint(100, 2);
        long el = Decimal64Utils.fromFixedPoint(1000, 3);

        Decimal64 a = Decimal64.fromUnderlying(al), b = Decimal64.fromUnderlying(bl), c = Decimal64.fromUnderlying(cl), d = Decimal64.fromUnderlying(dl), e = Decimal64.fromUnderlying(el);


        Assert.assertEquals(false, a.isIdentical(b));
        Assert.assertEquals(false, b.isIdentical(c));
        Assert.assertEquals(false, c.isIdentical(d));
        Assert.assertEquals(false, d.isIdentical(e));
        Assert.assertEquals(false, e.isIdentical(a));

        Assert.assertNotEquals(a.identityHashCode(), b.identityHashCode());
        Assert.assertNotEquals(b.identityHashCode(), c.identityHashCode());
        Assert.assertNotEquals(c.identityHashCode(), d.identityHashCode());
        Assert.assertNotEquals(d.identityHashCode(), e.identityHashCode());
        Assert.assertNotEquals(e.identityHashCode(), a.identityHashCode());

        Assert.assertEquals(true, a.equals(b));
        Assert.assertEquals(true, b.equals(c));
        Assert.assertEquals(true, c.equals(d));
        Assert.assertEquals(true, d.equals(e));
        Assert.assertEquals(true, e.equals(a));

        Assert.assertEquals(a.hashCode(), b.hashCode());
        Assert.assertEquals(b.hashCode(), c.hashCode());
        Assert.assertEquals(c.hashCode(), d.hashCode());
        Assert.assertEquals(d.hashCode(), e.hashCode());
        Assert.assertEquals(e.hashCode(), a.hashCode());

        a = a.canonize();
        b = b.canonize();
        c = c.canonize();
        d = d.canonize();
        e = e.canonize();
        Assert.assertTrue(a.isIdentical(b));
        Assert.assertTrue(a.isIdentical(c));
        Assert.assertTrue(a.isIdentical(d));
        Assert.assertTrue(a.isIdentical(e));

        long nan1l = Decimal64Utils.NaN;
        long nan2l = nan1l + 20;
        Decimal64 nan1 = Decimal64.fromUnderlying(nan1l), nan2 = Decimal64.fromUnderlying(nan2l);

        Assert.assertEquals(false, nan1.isIdentical(nan2));
        Assert.assertNotEquals(nan1.identityHashCode(), nan2.identityHashCode());
        Assert.assertEquals(true, nan1.equals(nan2));
        Assert.assertEquals(nan1.hashCode(), nan2.hashCode());

        nan1 = nan1.canonize();
        nan2 = nan2.canonize();
        Assert.assertTrue(nan1.isIdentical(nan2));

        long posInf1l = Decimal64Utils.POSITIVE_INFINITY;
        long posInf2l = posInf1l + 10;

        Decimal64 posInf1 = Decimal64.fromUnderlying(posInf1l), posInf2 = Decimal64.fromUnderlying(posInf2l);

        Assert.assertEquals(false, posInf1.isIdentical(posInf2));
        Assert.assertNotEquals(posInf1.identityHashCode(), posInf2.identityHashCode());
        Assert.assertEquals(true, posInf1.equals(posInf2));
        Assert.assertEquals(posInf1.hashCode(), posInf2.hashCode());


        posInf1 = posInf1.canonize();
        posInf2 = posInf2.canonize();

        Assert.assertEquals(true, posInf1.isIdentical(posInf2));

        long negInf1l = Decimal64Utils.NEGATIVE_INFINITY;
        long negInf2l = negInf1l + 10;

        Decimal64 negInf1 = Decimal64.fromUnderlying(negInf1l), negInf2 = Decimal64.fromUnderlying(negInf2l);

        Assert.assertEquals(false, negInf1.isIdentical(negInf2));
        Assert.assertNotEquals(negInf1.identityHashCode(), negInf2.identityHashCode());
        Assert.assertEquals(true, negInf1.equals(negInf2));
        Assert.assertEquals(negInf1.hashCode(), negInf2.hashCode());


        negInf1 = negInf1.canonize();
        negInf2 = negInf2.canonize();
        Assert.assertEquals(true, negInf1.isIdentical(negInf2));

        long zero1l = Decimal64Utils.fromFixedPoint(0, 1);
        long zero2l = Decimal64Utils.fromFixedPoint(0, 2);

        Decimal64 zero1 = Decimal64.fromUnderlying(zero1l);
        Decimal64 zero2 = Decimal64.fromUnderlying(zero2l);

        Assert.assertEquals(false, zero1.isIdentical(zero2));
        Assert.assertNotEquals(zero1.identityHashCode(), zero2.identityHashCode());
        Assert.assertEquals(true, zero1.equals(zero2));
        Assert.assertEquals(zero1.hashCode(), zero2.hashCode());

        zero1 = zero1.canonize();
        zero2 = zero2.canonize();
        Assert.assertEquals(zero1, zero2);
    }

    @Test
    public void numberConversionTest() {
        for (long i0 = 0; i0 < 9999999999999999L; i0 = (i0 << 1) + 1) {
            long x = i0;
            for (int i = 0; i < 2; i++, x = -x) {
                Assert.assertEquals(x, (long)Decimal64.fromDouble(x).toDouble());
                Assert.assertEquals(x, (long)Decimal64.fromDouble(x).toLong());
                Assert.assertTrue(Decimal64.fromDouble(x).equals(Decimal64.fromLong(x)));
                Assert.assertTrue(Decimal64.fromLong(x).equals(Decimal64.fromDouble(x)));

                if (Math.abs(x) <= Integer.MAX_VALUE) {
                    Assert.assertEquals(x, (long) Decimal64.fromLong(x).toInt());
                    Assert.assertEquals(x, (long) Decimal64.fromInt((int) x).toDouble());
                    Assert.assertEquals(x, (long) Decimal64.fromInt((int) x).toLong());
                    Assert.assertEquals(x, (long) Decimal64.fromInt((int) x).toInt());
                    Assert.assertTrue(Decimal64.fromDouble(x).equals(Decimal64.fromInt((int) x)));
                    Assert.assertTrue(Decimal64.fromLong(x).equals(Decimal64.fromInt((int) x)));
                } else {
                    // Expect integer overflow at some stage
                    Assert.assertNotEquals(x, (long) Decimal64.fromLong(x).toInt());
                    Assert.assertNotEquals(x, (long) Decimal64.fromInt((int) x).toDouble());
                    Assert.assertNotEquals(x, (long) Decimal64.fromInt((int) x).toLong());
                    Assert.assertNotEquals(x, (long) Decimal64.fromInt((int) x).toInt());
                    Assert.assertFalse(Decimal64.fromDouble(x).equals(Decimal64.fromInt((int) x)));
                    Assert.assertFalse(Decimal64.fromLong(x).equals(Decimal64.fromInt((int) x)));
                }
            }
        }
    }

    @Test
    public void toStringTest() {

        Assert.assertEquals("0", Decimal64.fromLong(0).toString());
        Assert.assertEquals("42", Decimal64.fromLong(42).toString());
        Assert.assertEquals(String.valueOf(Integer.MAX_VALUE), Decimal64.fromInt(Integer.MAX_VALUE).toString());
        Assert.assertEquals(String.valueOf(Integer.MIN_VALUE), Decimal64.fromInt(Integer.MIN_VALUE).toString());
        Assert.assertEquals(String.valueOf(Integer.MAX_VALUE), Decimal64.fromLong(Integer.MAX_VALUE).toString());
        Assert.assertEquals(String.valueOf(Integer.MIN_VALUE), Decimal64.fromLong(Integer.MIN_VALUE).toString());
        Assert.assertEquals(String.valueOf(Integer.MAX_VALUE), Decimal64.fromFixedPoint(Integer.MAX_VALUE, 0).toString());
        Assert.assertEquals(String.valueOf(Integer.MIN_VALUE), Decimal64.fromFixedPoint(Integer.MIN_VALUE, 0).toString());
        Assert.assertEquals("123.456", Decimal64.fromDouble(123.456).toString());
        Assert.assertEquals("123.4567", Decimal64.fromFixedPoint(1234567, 4).toString());

        Assert.assertEquals("null", String.valueOf((Object)null));
        Assert.assertEquals("null", String.valueOf((Decimal64)null));
        Assert.assertEquals("null", Decimal64.toString(null));
        Assert.assertEquals("null", Decimal64Utils.toString(Decimal64.toUnderlying((Decimal64)null)));

        Assert.assertEquals("NaN", String.valueOf(Decimal64.NaN));
        Assert.assertEquals("NaN", Decimal64.toString(Decimal64.NaN.negate()));
        Assert.assertEquals("Infinity", Decimal64.toString(Decimal64.POSITIVE_INFINITY));
        Assert.assertEquals("-Infinity", Decimal64.toString(Decimal64.NEGATIVE_INFINITY));

        Assert.assertEquals("0", Decimal64.toString(Decimal64.ZERO));
        Assert.assertEquals("1000000", Decimal64.toString(Decimal64.MILLION));
        Assert.assertEquals("0.01", Decimal64.toString(Decimal64.ONE_HUNDREDTH));
    }
}

File: csharp/Deltix.DFP.Test/Decimal64Test.cs
using System;
using System.IO;
using System.Runtime.Serialization.Formatters.Binary;
using NUnit.Framework;

namespace Deltix.DFP.Test
{
	[TestFixture]
	class Decimal64Test
	{
		[TestCase()]
		public void IsNullTest()
		{
			Decimal64 a = Decimal64.Null;
			Assert.AreEqual(true, a.IsNull());
			Assert.AreEqual(true, a.IsNaN());
			Decimal64 b = Decimal64.NaN;
			Assert.AreEqual(false, b.IsNull());
			Assert.AreEqual(true, b.IsNaN());
		}

		[TestCase()]
		public void CanonizeTest()
		{
			Decimal64 a = Decimal64.FromFixedPoint(10000, 4);
			Decimal64 b = Decimal64.FromFixedPoint(1, 0);
			Decimal64 c = Decimal64.FromFixedPoint(10, 1);
			Decimal64 d = Decimal64.FromFixedPoint(100, 2);
			Decimal64 e = Decimal64.FromFixedPoint(1000, 3);

			Assert.AreEqual(true, a.Equals(b));
			Assert.AreEqual(true, b.Equals(c));
			Assert.AreEqual(true, c.Equals(d));
			Assert.AreEqual(true, d.Equals(e));
			Assert.AreEqual(true, e.Equals(a));

			Assert.AreEqual(a.GetHashCode(), b.GetHashCode());
			Assert.AreEqual(b.GetHashCode(), c.GetHashCode());
			Assert.AreEqual(c.GetHashCode(), d.GetHashCode());
			Assert.AreEqual(d.GetHashCode(), e.GetHashCode());
			Assert.AreEqual(e.GetHashCode(), a.GetHashCode());

			Decimal64 nan1 = Decimal64.NaN;
			Decimal64 nan2 = Decimal64.FromUnderlying(nan1.Bits + 20);

			Assert.AreEqual(true, nan1.Equals(nan2));
			Assert.AreEqual(nan1.GetHashCode(), nan2.GetHashCode());


			Decimal64 posInf1 = Decimal64.PositiveInfinity;
			Decimal64 posInf2 = Decimal64.FromUnderlying(posInf1.Bits + 10);

			Assert.AreEqual(true, posInf1.Equals(posInf2));
			Assert.AreEqual(posInf1.GetHashCode(), posInf2.GetHashCode());

			Decimal64 negInf1 = Decimal64.PositiveInfinity;
			Decimal64 negInf2 = Decimal64.FromUnderlying(negInf1.Bits + 10);

			Assert.AreEqual(true, negInf1.Equals(negInf2));
			Assert.AreEqual(negInf1.GetHashCode(), negInf2.GetHashCode());

			Decimal64 zero1 = Decimal64.FromFixedPoint(0, 1);
			Decimal64 zero2 = Decimal64.FromFixedPoint(0, 2);

			Assert.AreEqual(true, zero1.Equals(zero2));
			Assert.AreEqual(zero1.GetHashCode(), zero2.GetHashCode());

		}


		[Test]
		public void Binary64Conversion()
		{
			Double binary64 = new Random().NextDouble();
			Decimal64 decimal64 = Decimal64.FromDouble(binary64);
			Double result = (Double)decimal64;
			Assert.That(Math.Abs(binary64 - result), Is.LessThanOrEqualTo(1.0E-16));
		}

		[Test]
		public void FromFixedPointFastConsts()
		{
			TestUtils.AssertDfp(Decimal64.Zero.Bits, DotNetImpl.FromFixedPointFast(0, 0));
			AssertDfp(Decimal64.One, DotNetImpl.FromFixedPointFast(1, 0));
			AssertDfp(Decimal64.Two, DotNetImpl.FromFixedPointFast(2, 0));
			AssertDfp(Decimal64.Ten, DotNetImpl.FromFixedPointFast(10, 0));
			AssertDfpEq(Decimal64.Ten, DotNetImpl.FromFixedPointFast(1, -1));
			AssertDfp(Decimal64.Hundred, DotNetImpl.FromFixedPointFast(100, 0));
			AssertDfpEq(Decimal64.Hundred, DotNetImpl.FromFixedPointFast(1, -2));
			AssertDfpEq(Decimal64.Thousand, DotNetImpl.FromFixedPointFast(1, -3));
			AssertDfpEq(Decimal64.Million, DotNetImpl.FromFixedPointFast(1, -6));
			AssertDfp(Decimal64.OneTenth, DotNetImpl.FromFixedPointFast(1, 1));
			AssertDfp(Decimal64.OneHundredth, DotNetImpl.FromFixedPointFast(1, 2));

			AssertDfp(Decimal64.Zero, DotNetImpl.FromFixedPointFastUnsigned(0, 0));
			AssertDfp(Decimal64.One, DotNetImpl.FromFixedPointFastUnsigned(1, 0));
			AssertDfp(Decimal64.Two, DotNetImpl.FromFixedPointFastUnsigned(2, 0));
			AssertDfp(Decimal64.Ten, DotNetImpl.FromFixedPointFastUnsigned(10, 0));
			AssertDfpEq(Decimal64.Ten, DotNetImpl.FromFixedPointFastUnsigned(1, -1));
			AssertDfp(Decimal64.Hundred, DotNetImpl.FromFixedPointFastUnsigned(100, 0));
			AssertDfpEq(Decimal64.Hundred, DotNetImpl.FromFixedPointFastUnsigned(1, -2));
			AssertDfpEq(Decimal64.Thousand, DotNetImpl.FromFixedPointFastUnsigned(1, -3));
			AssertDfpEq(Decimal64.Million, DotNetImpl.FromFixedPointFastUnsigned(1, -6));
		}

		private void AssertDfp(UInt64 actual, UInt64 expected) => TestUtils.AssertDfp(actual, expected);
		private void AssertDfp(UInt64 actual, Decimal64 expected) => TestUtils.AssertDfp(actual, expected.Bits);
		private void AssertDfp(Decimal64 actual, Decimal64 expected) => TestUtils.AssertDfp(actual, expected);
		private void AssertDfp(Decimal64 actual, UInt64 expected) => TestUtils.AssertDfp(actual.Bits, expected);
		private void AssertDfpEq(Decimal64 actual, Decimal64 expected) => TestUtils.AssertDfp(actual, expected);
		private void AssertDfpEq(Decimal64 actual, UInt64 expected) => TestUtils.AssertDfpEq(actual, Decimal64.FromUnderlying(expected));
		

		[Test]
		public void FromFixedPointFast()
		{
			int N = 10000;
			Random random = new Random();
			for (int exp = 398 - 0x2FF; exp <= 398; ++exp)
			{
				for (int j = 0; j < N; ++j)
				{
					int mantissa = random.Next();
					long mantissa64 = ((long)mantissa << 32) + random.Next();
					if (random.Next() < 0)
						mantissa64 = -mantissa64;

					UInt64 correct32 = NativeImpl.fromFixedPoint32(mantissa, exp);
					UInt64 correctU32 = NativeImpl.fromFixedPointU32((UInt32)mantissa, exp);
					UInt64 correct64 = NativeImpl.fromFixedPoint64(mantissa, exp);

					AssertDfp(correct32, DotNetImpl.FromFixedPointFast(mantissa, exp));
					AssertDfp(correct64, DotNetImpl.FromFixedPointFast(mantissa, exp));
					AssertDfp(correct64, DotNetImpl.FromFixedPointFast(mantissa, exp));

					AssertDfp(correctU32, DotNetImpl.FromFixedPointFastUnsigned((UInt32)mantissa, exp));
					AssertDfp(NativeImpl.fromFixedPoint64((UInt32)mantissa, exp), DotNetImpl.FromFixedPointFastUnsigned((UInt32)mantissa, exp));

					AssertDfp(correct32, Decimal64.FromFixedPoint(mantissa, exp));
					AssertDfp(correct64, Decimal64.FromFixedPoint(mantissa, exp));
					AssertDfp(correctU32, Decimal64.FromFixedPoint((UInt32)mantissa, exp));
					AssertDfp(NativeImpl.fromFixedPoint64((UInt32)mantissa, exp), Decimal64.FromFixedPoint((UInt32)mantissa, exp));
					AssertDfp(correct32, Decimal64.FromFixedPoint((Int64)mantissa, exp));
					AssertDfp(correct64, Decimal64.FromFixedPoint((Int64)mantissa, exp));

					if (mantissa >= 0)
					{
						AssertDfp(correct32, DotNetImpl.FromFixedPointFastUnsigned((UInt32)mantissa, exp));
						AssertDfp(correct64, DotNetImpl.FromFixedPointFastUnsigned((UInt32)mantissa, exp));
					}

					for (int k = 0; k < 32; ++k)
					{
						AssertDfp(NativeImpl.fromFixedPoint64(mantissa, exp), Decimal64.FromFixedPoint(mantissa, exp));
						AssertDfp(NativeImpl.fromFixedPoint64(mantissa64, exp), Decimal64.FromFixedPoint(mantissa64, exp));
						mantissa >>= 1;
						mantissa64 >>= 1;
					}
				}

				AssertDfp(NativeImpl.fromFixedPoint32(0, exp), DotNetImpl.FromFixedPointFast(0, exp));
				AssertDfp(NativeImpl.fromFixedPoint32(Int32.MinValue, exp), DotNetImpl.FromFixedPointFast(Int32.MinValue, exp));
				AssertDfp(NativeImpl.fromFixedPoint32(Int32.MaxValue, exp), DotNetImpl.FromFixedPointFast(Int32.MaxValue, exp));
				AssertDfp(NativeImpl.fromFixedPoint64(0, exp), DotNetImpl.FromFixedPointFast(0, exp));
				AssertDfp(NativeImpl.fromFixedPoint64(Int32.MinValue, exp), DotNetImpl.FromFixedPointFast(Int32.MinValue, exp));
				AssertDfp(NativeImpl.fromFixedPoint64(Int32.MaxValue, exp), DotNetImpl.FromFixedPointFast(Int32.MaxValue, exp));
			}
		}

		[Test]
		public void FromFixedPointFastMin()
		{
			Assert.Throws<ArgumentException>(delegate { DotNetImpl.FromFixedPointFast(0, 398 - 0x300); });
		}

		[Test]
		public void FromFixedPointFastMax()
		{
			Assert.Throws<ArgumentException>(delegate { DotNetImpl.FromFixedPointFast(0, 399); });
		}

		[Test]
		public void FromFixedPointFastUMin()
		{
			Assert.Throws<ArgumentException>(delegate { DotNetImpl.FromFixedPointFastUnsigned(0, 398 - 0x300); });
		}

		[Test]
		public void FromFixedPointFastUMax()
		{
			Assert.Throws<ArgumentException>(delegate { DotNetImpl.FromFixedPointFastUnsigned(0, 399); });
		}


		private String PrintFp(Object a) => a.ToString().Replace(',', '.');

		private void AssertDecimalEqual(Object a, Object b) => Assert.AreEqual(PrintFp(a), PrintFp(b));
		private void AssertDecimalNotEqual(Object a, Object b) => Assert.AreNotEqual(PrintFp(a), PrintFp(b));

		[Test]
		public void DecimalInternalRepresentation()
		{
			Random random = new Random();
			Assert.AreEqual(16, sizeof(Decimal));
			for (int i = 0; i < 1000; ++i)
			{
				int lo = random.Next(), mid = random.Next(), hi = random.Next();
				byte exp = (byte)random.Next(0, 28);
				bool sign = random.Next() < 0;
				var x = new Decimal(lo, mid, hi, sign, exp);
				int lo2, mid2, hi2;
				byte exp2;
				bool sign2;
				UInt64 lo64;
				unsafe
				{
					int * f = (int *)&x; // Order (Little endian): flags, hi, lo, mid
					lo64 = ((ulong*) &x)[1];
					lo2 = f[2];
					mid2 = f[3];
					hi2 = f[1];
					short flg16 = ((short*) &x)[1];
					sign2 = flg16 < 0;
					//exp2 = (byte) (f[0] >> 16);
					exp2 = (byte)flg16;
				}

				Assert.AreEqual(lo, lo2);
				Assert.AreEqual(mid, mid2);
				Assert.AreEqual((uint)lo | ((ulong)(uint)mid << 32), lo64);
				Assert.AreEqual(lo, lo2);
				Assert.AreEqual(hi, hi2);
				Assert.AreEqual(sign, sign2);
				Assert.AreEqual(exp, exp2);
			}
		}


		[Test]
		public void DecimalConversionBasicFrom()
		{
			Decimal x = new Decimal(6) / new Decimal(10000);
			var bits = Decimal.GetBits(x);


			AssertDecimalEqual(x, (Decimal64)x);
			AssertDecimalEqual(x, (Decimal64)x);
			AssertDecimalEqual(x, Decimal64.FromDecimal(x));
			AssertDecimalEqual(0.0006M, (Decimal64)0.0006M);
			AssertDecimalEqual(0.0000006M, (Decimal64)0.0000006M);
			AssertDecimalEqual(0.00000000000006M, (Decimal64)0.00000000000006M);
			AssertDecimalEqual(9.8765432198721M, (Decimal64)9.8765432198721M);
			AssertDecimalEqual(0.00000098765432198721M, (Decimal64)0.00000098765432198721M);
			AssertDecimalEqual(0.0000000000000098765432198721M, (Decimal64)0.0000000000000098765432198721M);
			AssertDecimalEqual(1234567890123456M, (Decimal64)1234567890123456M);
			AssertDecimalEqual(1234567890123456000M, (Decimal64)1234567890123456000M);
			AssertDecimalEqual(0.00000000001M, (Decimal64)0.00000000001M);
			AssertDecimalNotEqual(1234567890123456111111M, (Decimal64)123456789012345111111M);

			AssertDecimalNotEqual(0.0006M, (Decimal64)(double)0.0006M);
			AssertDecimalNotEqual(0.0000000000000098765432198721M, (Decimal64)(double)0.0000000000000098765432198721M);
			AssertDecimalNotEqual(1234567890123456111111M, (Decimal64)(double)123456789012345111111M);
		}

		[Test]
		public void DecimalConversionTo()
		{
			Decimal x = new Decimal(6) / new Decimal(10000);
			Assert.AreEqual(x, (Decimal)Decimal64.FromFixedPoint(6, 4));
			AssertDecimalEqual(1234567890123456M, (Decimal)Decimal64.FromLong(1234567890123456));
			AssertDecimalEqual(0.1234567890123456M, (Decimal)Decimal64.FromFixedPoint(1234567890123456, 16));
			AssertDecimalEqual(0.00000000001234567890123456M, (Decimal)Decimal64.FromFixedPoint(1234567890123456, 26));

			AssertDecimalEqual(123456789012345.6M, (Decimal)Decimal64.FromFixedPoint(1234567890123456, 1));
			Assert.AreEqual(123456789012345.600000000M, (Decimal)(Decimal64.FromFixedPoint(1234567890123456, 1)));
			Assert.AreEqual(123456789012300.000000000M, (Decimal)(Decimal64.FromFixedPoint(1234567890123000, 1)));
			Assert.AreEqual(123456789012300.000000000M, (Decimal)(Decimal64.FromFixedPoint(1234567890123, -2)));

			// These are not converted precisely(TODO:)
			AssertDecimalNotEqual(12345678901234560M, (Decimal)Decimal64.FromLong(12345678901234560));
			AssertDecimalNotEqual(1234567890123456000M, (Decimal)Decimal64.FromLong(1234567890123456000));
			AssertDecimalNotEqual(123456789012345600000000M, (Decimal)Decimal64.FromFixedPoint(1234567890123456, -8));
			AssertDecimalNotEqual(7345678901234560001M, (Decimal)Decimal64.FromLong(7345678901234560001));
		}

		[Test]
		public void DecimalConversion2()
		{
			Decimal x = new Decimal(6) / new Decimal(10000);
			Assert.AreEqual(x, (Decimal)(Decimal64)x);
		}
		

		[Test]
		public void NumberConversionTest()
		{
			for (long i0 = 0; i0 < 9999999999999999L; i0 = (i0 << 1) + 1)
			{
				long x = i0;
				for (int i = 0; i < 2; i++, x = -x)
				{
					Assert.That(x, Is.EqualTo((long)Decimal64.FromDouble(x).ToDouble()));
					Assert.That(x, Is.EqualTo((long)Decimal64.FromDouble(x).ToLong()));
					Assert.True(Decimal64.FromDouble(x).Equals(Decimal64.FromLong(x)));
					Assert.True(Decimal64.FromLong(x).Equals(Decimal64.FromDouble(x)));
					Assert.That(x, Is.EqualTo((long)Decimal64.FromDouble(x)));

					if (x >= 0)
					{
						Assert.That(x, Is.EqualTo((long)(UInt64)Decimal64.FromDouble(x)));
					}
					else
					{
						Assert.True(1UL << 31 == (UInt32)Decimal64.FromLong(x));
						Assert.True(1UL << 63 == (UInt64)Decimal64.FromLong(x));
					}

					if (Math.Abs(x) <= Int32.MaxValue)
					{
						Assert.That(x, Is.EqualTo((long)Decimal64.FromLong(x).ToInt()));
						Assert.That(x, Is.EqualTo((long)Decimal64.FromInt((int)x).ToDouble()));
						Assert.That(x, Is.EqualTo((long)Decimal64.FromInt((int)x).ToLong()));
						Assert.That(x, Is.EqualTo((long)Decimal64.FromInt((int)x).ToInt()));
						Assert.True(Decimal64.FromDouble(x).Equals(Decimal64.FromInt((int)x)));
						Assert.True(Decimal64.FromLong(x).Equals(Decimal64.FromInt((int)x)));

						Assert.That(x, Is.EqualTo((int)Decimal64.FromDouble(x)));

						if (x >= 0)
						{
							Assert.That(x, Is.EqualTo((UInt32)Decimal64.FromDouble(x)));
						}
						else
						{
							Assert.That(1UL << 31, Is.EqualTo((UInt32)Decimal64.FromDouble(x)));
						}
					}
					else
					{
						// Expect integer overflow at some stage
						Assert.That(x, Is.Not.EqualTo((long)Decimal64.FromLong(x).ToInt()));
						Assert.That(x, Is.Not.EqualTo((long)Decimal64.FromInt((int)x).ToDouble()));
						Assert.That(x, Is.Not.EqualTo((long)Decimal64.FromInt((int)x).ToLong()));
						Assert.That(x, Is.Not.EqualTo((long)Decimal64.FromInt((int)x).ToInt()));
						Assert.False(Decimal64.FromDouble(x).Equals(Decimal64.FromInt((int)x)));
						Assert.False(Decimal64.FromLong(x).Equals(Decimal64.FromInt((int)x)));
					}
				}
			}
		}

		[Test]
		public void Min()
		{
			Assert.That(Decimal64.NaN.Min(Decimal64.Zero).Bits, Is.EqualTo(Decimal64.NaN.Bits));
			Assert.That(Decimal64.Zero.Min(Decimal64.NaN).Bits, Is.EqualTo(Decimal64.NaN.Bits));
			Assert.That(Decimal64.NaN.Min(Decimal64.NaN).Bits, Is.EqualTo(Decimal64.NaN.Bits));
		}

		[Test]
		public void Max()
		{
			Assert.That(Decimal64.NaN.Max(Decimal64.Zero).Bits, Is.EqualTo(Decimal64.NaN.Bits));
			Assert.That(Decimal64.Zero.Max(Decimal64.NaN).Bits, Is.EqualTo(Decimal64.NaN.Bits));
			Assert.That(Decimal64.NaN.Max(Decimal64.NaN).Bits, Is.EqualTo(Decimal64.NaN.Bits));
		}

		[Test]
		public void CompareTo()
		{
			Assert.That(Decimal64.FromDouble(1.0).CompareTo(Decimal64.Zero), Is.EqualTo(1.0.CompareTo(0.0)));
			Assert.That(Decimal64.Zero.CompareTo(Decimal64.Zero), Is.EqualTo(0.0.CompareTo(0.0)));
			Assert.That(Decimal64.Zero.CompareTo(Decimal64.FromDouble(1.0)), Is.EqualTo(0.0.CompareTo(1.0)));

			Assert.That(Decimal64.NaN.CompareTo(Decimal64.Zero), Is.EqualTo(Double.NaN.CompareTo(0.0)));
			Assert.That(Decimal64.Zero.CompareTo(Decimal64.NaN), Is.EqualTo(0.0.CompareTo(Double.NaN)));
			Assert.That(Decimal64.NaN.CompareTo(Decimal64.NaN), Is.EqualTo(Double.NaN.CompareTo(Double.NaN)));
		}

		[Test]
		public void TestSerializable()
		{
			var a1 = Decimal64.FromDouble(123.45);
			var b1 = Decimal64.FromLong(42);

			var formatter = new BinaryFormatter();
			var stream = new MemoryStream();
			formatter.Serialize(stream, a1);
			for (int i = 0; i < 40; ++i)
			{
				formatter.Serialize(stream, b1);
			}

			formatter.Serialize(stream, a1);

			//String l = stream.Length.ToString();
			//stream.Seek(0, SeekOrigin.Begin);
			//var fs = new FileStream("d:\\serialized.bin", FileMode.Create, FileAccess.Write);
			//stream.WriteTo(fs);
			//fs.Close();

			stream.Seek(0, SeekOrigin.Begin);

			Assert.True(a1.Equals(formatter.Deserialize(stream)));
			for (int i = 0; i < 40; ++i)
			{
				Assert.True(b1.Equals(formatter.Deserialize(stream)));
			}

			Assert.True(a1.Equals(formatter.Deserialize(stream)));
			stream.Close();
		}

		[Test]
		public void TestFromDecimalDoubleBasic()
		{
			Assert.AreNotEqual("9.2", Decimal64.FromDouble(9.2).ToString());
			Assert.AreEqual("9.199999999999999", Decimal64.FromDouble(9.2).ToString());
			Assert.AreEqual("9.2", Decimal64.FromDecimalDouble(9.2).ToString());
		}

		static void Main(string[] args)
		{
		}
	}
}
